File: selenium-tests/helpers/excelReporter.js
// ============================================================
// Excel Report Generator — Selenium Tests
// ============================================================
'use strict';

const ExcelJS  = require('exceljs');
const path     = require('path');
const fs       = require('fs');
const config   = require('../config/testConfig');

// ── Colour palette ──────────────────────────────────────────
const COLORS = {
  PASS:       { argb: 'FF4CAF50' },   // Green
  FAIL:       { argb: 'FFF44336' },   // Red
  SKIP:       { argb: 'FFFFC107' },   // Amber
  HEADER_BG:  { argb: 'FF1A1A2E' },   // Dark navy
  HEADER_FG:  { argb: 'FFFFFFFF' },   // White
  ALT_ROW:    { argb: 'FFF5F5F5' },   // Light grey
  TITLE_BG:   { argb: 'FF16213E' },   // Darker navy
  ACCENT:     { argb: 'FFE94560' }    // Coral accent
};

const FONT_HEADER = { name: 'Calibri', size: 11, bold: true, color: COLORS.HEADER_FG };
const FONT_TITLE  = { name: 'Calibri', size: 16, bold: true, color: COLORS.HEADER_FG };
const FONT_BODY   = { name: 'Calibri', size: 10 };
const FONT_PASS   = { name: 'Calibri', size: 10, bold: true, color: { argb: 'FF2E7D32' } };
const FONT_FAIL   = { name: 'Calibri', size: 10, bold: true, color: { argb: 'FFC62828' } };

/**
 * Generate a full Excel test report.
 * @param {Array} results  - Array of test result objects
 * @param {Object} summary - { total, passed, failed, skipped, duration, suite }
 */
async function generateReport(results, summary) {
  if (!fs.existsSync(config.reportDir)) {
    fs.mkdirSync(config.reportDir, { recursive: true });
  }

  const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
  const filename  = `${config.reportName}_${timestamp}.xlsx`;
  const filepath  = path.join(config.reportDir, filename);

  const workbook = new ExcelJS.Workbook();
  workbook.creator  = 'FoodBridge Selenium Suite';
  workbook.created  = new Date();
  workbook.modified = new Date();

  _buildSummarySheet(workbook, summary, results);
  _buildDetailsSheet(workbook, results);
  _buildCategorySheet(workbook, results);

  await workbook.xlsx.writeFile(filepath);
  console.log(`\n  📊 Excel report saved → ${filepath}\n`);
  return filepath;
}

// ── Summary Sheet ───────────────────────────────────────────
function _buildSummarySheet(workbook, summary, results) {
  const ws = workbook.addWorksheet('📋 Summary', {
    properties: { tabColor: { argb: 'FF1A1A2E' } }
  });

  ws.columns = [
    { width: 30 }, { width: 25 }, { width: 20 }, { width: 20 }
  ];

  // Title
  ws.mergeCells('A1:D1');
  const titleCell = ws.getCell('A1');
  titleCell.value = '🍱 FoodBridge Web Application — Test Report';
  titleCell.font = FONT_TITLE;
  titleCell.fill = { type: 'pattern', pattern: 'solid', fgColor: COLORS.TITLE_BG };
  titleCell.alignment = { horizontal: 'center', vertical: 'middle' };
  ws.getRow(1).height = 36;

  // Metadata
  const meta = [
    ['Generated At',   new Date().toLocaleString()],
    ['Suite Name',     summary.suite || 'FoodBridge Web E2E'],
    ['Browser',        config.browser.toUpperCase()],
    ['Base URL',       config.baseUrl],
    ['Total Duration', `${(summary.duration / 1000).toFixed(2)}s`]
  ];

  meta.forEach(([k, v], i) => {
    const row = ws.getRow(i + 2);
    row.getCell(1).value = k;
    row.getCell(2).value = v;
    row.getCell(1).font  = { bold: true, name: 'Calibri' };
    row.getCell(2).font  = FONT_BODY;
    row.height = 18;
  });

  ws.addRow([]);

  // Stats cards
  const statsRow = ws.getRow(8);
  const stats = [
    { label: 'Total Tests',   value: summary.total,   color: 'FF1A1A2E' },
    { label: '✅ Passed',     value: summary.passed,  color: 'FF4CAF50' },
    { label: '❌ Failed',     value: summary.failed,  color: 'FFF44336' },
    { label: '⏭ Skipped',    value: summary.skipped, color: 'FFFFC107' }
  ];
  stats.forEach((s, i) => {
    const col = i + 1;
    ws.mergeCells(8, col, 9, col);
    const cell = ws.getCell(8, col);
    cell.value     = `${s.label}\n${s.value}`;
    cell.font      = { name: 'Calibri', size: 14, bold: true, color: { argb: 'FFFFFFFF' } };
    cell.fill      = { type: 'pattern', pattern: 'solid', fgColor: { argb: s.color } };
    cell.alignment = { horizontal: 'center', vertical: 'middle', wrapText: true };
    ws.getRow(8).height = 40;
  });

  ws.addRow([]);
  ws.addRow([]);

  // Pass rate bar
  const passRate = summary.total > 0
    ? ((summary.passed / summary.total) * 100).toFixed(1)
    : 0;
  const rateRow = ws.addRow([`Pass Rate: ${passRate}%`]);
  rateRow.getCell(1).font = { bold: true, size: 12, name: 'Calibri', color: { argb: 'FF2E7D32' } };

  ws.addRow([]);

  // Category breakdown header
  const catHeader = ws.addRow(['Category', 'Total', 'Passed', 'Failed', 'Pass %']);
  catHeader.eachCell(cell => {
    cell.font = FONT_HEADER;
    cell.fill = { type: 'pattern', pattern: 'solid', fgColor: COLORS.HEADER_BG };
    cell.alignment = { horizontal: 'center' };
  });
  ws.getRow(catHeader.number).height = 22;

  // Category breakdown data
  const byCategory = {};
  results.forEach(r => {
    if (!byCategory[r.category]) byCategory[r.category] = { total: 0, passed: 0, failed: 0 };
    byCategory[r.category].total++;
    if (r.status === 'PASS') byCategory[r.category].passed++;
    else if (r.status === 'FAIL') byCategory[r.category].failed++;
  });

  Object.entries(byCategory).forEach(([cat, data], i) => {
    const pct = data.total > 0 ? ((data.passed / data.total) * 100).toFixed(0) : 0;
    const row = ws.addRow([cat, data.total, data.passed, data.failed, `${pct}%`]);
    if (i % 2 === 0) {
      row.eachCell(cell => {
        cell.fill = { type: 'pattern', pattern: 'solid', fgColor: COLORS.ALT_ROW };
      });
    }
    row.getCell(3).font = { color: { argb: 'FF2E7D32' }, bold: true, name: 'Calibri' };
    row.getCell(4).font = { color: { argb: 'FFC62828' }, bold: true, name: 'Calibri' };
    row.eachCell(cell => { cell.alignment = { horizontal: 'center' }; });
  });
}

// ── Test Details Sheet ──────────────────────────────────────
function _buildDetailsSheet(workbook, results) {
  const ws = workbook.addWorksheet('📝 Test Details', {
    properties: { tabColor: { argb: 'FF4CAF50' } }
  });

  ws.columns = [
    { header: 'Test ID',      key: 'id',       width: 10 },
    { header: 'Category',     key: 'category', width: 25 },
    { header: 'Test Name',    key: 'name',     width: 50 },
    { header: 'Type',         key: 'type',     width: 18 },
    { header: 'Status',       key: 'status',   width: 12 },
    { header: 'Duration (s)', key: 'duration', width: 15 },
    { header: 'Error / Notes', key: 'error',  width: 55 }
  ];

  // Style header row
  const headerRow = ws.getRow(1);
  headerRow.eachCell(cell => {
    cell.font      = FONT_HEADER;
    cell.fill      = { type: 'pattern', pattern: 'solid', fgColor: COLORS.HEADER_BG };
    cell.alignment = { horizontal: 'center', vertical: 'middle' };
    cell.border    = {
      bottom: { style: 'medium', color: COLORS.ACCENT }
    };
  });
  headerRow.height = 24;

  // Data rows
  results.forEach((r, i) => {
    const row = ws.addRow({
      id:       r.id,
      category: r.category,
      name:     r.name,
      type:     r.type || 'Functional',
      status:   r.status,
      duration: r.duration ? (r.duration / 1000).toFixed(2) : '-',
      error:    r.error || ''
    });

    // Alternate row shading
    if (i % 2 === 0) {
      row.eachCell(cell => {
        cell.fill = { type: 'pattern', pattern: 'solid', fgColor: COLORS.ALT_ROW };
      });
    }

    // Status cell colouring
    const statusCell = row.getCell('status');
    if (r.status === 'PASS') {
      statusCell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFBBDEFB' } };
      statusCell.font = FONT_PASS;
    } else if (r.status === 'FAIL') {
      statusCell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFFFCDD2' } };
      statusCell.font = FONT_FAIL;
    } else {
      statusCell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFFFF9C4' } };
      statusCell.font = { name: 'Calibri', size: 10, bold: true, color: { argb: 'FFE65100' } };
    }

    statusCell.alignment = { horizontal: 'center' };
    row.getCell('error').alignment = { wrapText: true };
    row.height = 20;
  });

  // Auto filter
  ws.autoFilter = { from: 'A1', to: 'G1' };
  ws.views = [{ state: 'frozen', ySplit: 1 }];
}

// ── Category Sheet ──────────────────────────────────────────
function _buildCategorySheet(workbook, results) {
  const ws = workbook.addWorksheet('📊 By Category', {
    properties: { tabColor: { argb: 'FFE94560' } }
  });

  ws.columns = [
    { header: 'Category',  key: 'category', width: 30 },
    { header: 'Test Type', key: 'type',     width: 20 },
    { header: 'Test Name', key: 'name',     width: 50 },
    { header: 'Status',    key: 'status',   width: 12 }
  ];

  const headerRow = ws.getRow(1);
  headerRow.eachCell(cell => {
    cell.font  = FONT_HEADER;
    cell.fill  = { type: 'pattern', pattern: 'solid', fgColor: COLORS.HEADER_BG };
    cell.alignment = { horizontal: 'center' };
  });
  headerRow.height = 22;

  // Sort by category
  const sorted = [...results].sort((a, b) =>
    a.category.localeCompare(b.category)
  );

  sorted.forEach((r, i) => {
    const row = ws.addRow({
      category: r.category,
      type:     r.type || 'Functional',
      name:     r.name,
      status:   r.status
    });

    if (i % 2 === 0) {
      row.eachCell(cell => {
        cell.fill = { type: 'pattern', pattern: 'solid', fgColor: COLORS.ALT_ROW };
      });
    }

    const sc = row.getCell('status');
    if (r.status === 'PASS') {
      sc.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFBBDEFB' } };
      sc.font = FONT_PASS;
    } else if (r.status === 'FAIL') {
      sc.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFFFCDD2' } };
      sc.font = FONT_FAIL;
    }
    sc.alignment = { horizontal: 'center' };
  });

  ws.autoFilter = { from: 'A1', to: 'D1' };
  ws.views = [{ state: 'frozen', ySplit: 1 }];
}

module.exports = { generateReport };

File: appium-python-tests/helpers/excel_reporter.py
# ============================================================
# Excel Report Generator — Appium Tests (Python)
# ============================================================
import os
from datetime import datetime
from openpyxl import Workbook
from openpyxl.styles import Font, PatternFill, Alignment, Side, Border
from openpyxl.utils import get_column_letter
from config.appium_config import CONFIG

COLORS = {
    "PASS": "4CAF50",
    "FAIL": "F44336",
    "SKIP": "FFC107",
    "HEADER_BG": "0D3349",
    "HEADER_FG": "FFFFFF",
    "ALT_ROW": "F5F5F5",
    "TITLE_BG": "0A2740"
}

FONT_NAME = "Calibri"
FONT_HEADER = Font(name=FONT_NAME, size=11, bold=True, color=COLORS["HEADER_FG"])
FONT_TITLE = Font(name=FONT_NAME, size=16, bold=True, color=COLORS["HEADER_FG"])
FONT_PASS = Font(name=FONT_NAME, size=10, bold=True, color="2E7D32")
FONT_FAIL = Font(name=FONT_NAME, size=10, bold=True, color="C62828")
FONT_BOLD = Font(name=FONT_NAME, size=10, bold=True)
FONT_BODY = Font(name=FONT_NAME, size=10)

def generate_report(results, summary):
    report_dir = CONFIG["report_dir"]
    if not os.path.exists(report_dir):
        os.makedirs(report_dir, exist_ok=True)

    timestamp = datetime.now().strftime("%Y-%m-%dT%H-%M-%S")
    filename = f"{CONFIG['report_name']}_{timestamp}.xlsx"
    filepath = os.path.join(report_dir, filename)

    wb = Workbook()
    
    # Remove default sheet
    wb.remove(wb.active)

    # Build sheets
    _build_summary_sheet(wb, summary, results)
    _build_details_sheet(wb, results)
    _build_category_sheet(wb, results)

    wb.save(filepath)
    print(f"\n  [Report] Mobile Excel report saved -> {filepath}\n")
    return filepath

def _build_summary_sheet(wb, summary, results):
    ws = wb.create_sheet(title="📋 Summary")
    ws.sheet_properties.tabColor = COLORS["HEADER_BG"]

    # Set Column widths
    ws.column_dimensions["A"].width = 30
    ws.column_dimensions["B"].width = 25
    ws.column_dimensions["C"].width = 20
    ws.column_dimensions["D"].width = 20

    # Title
    ws.merge_cells("A1:D1")
    title_cell = ws["A1"]
    title_cell.value = "📱 FoodBridge Android App — Test Report"
    title_cell.font = FONT_TITLE
    title_cell.fill = PatternFill(start_color=COLORS["TITLE_BG"], end_color=COLORS["TITLE_BG"], fill_type="solid")
    title_cell.alignment = Alignment(horizontal="center", vertical="center")
    ws.row_dimensions[1].height = 36

    # Metadata
    meta = [
        ("Generated At", datetime.now().strftime("%d/%m/%Y, %H:%M:%S")),
        ("Platform", "Android"),
        ("Device", CONFIG["capabilities"]["appium:deviceName"]),
        ("Android Version", CONFIG["capabilities"]["appium:platformVersion"]),
        ("Total Duration", f"{(summary['duration'] / 1000):.2f}s")
    ]
    for idx, (k, v) in enumerate(meta):
        row_num = idx + 2
        ws.cell(row=row_num, column=1, value=k).font = FONT_BOLD
        ws.cell(row=row_num, column=2, value=v).font = FONT_BODY
        ws.row_dimensions[row_num].height = 18

    # Stats cards
    stats = [
        {"label": "Total Tests", "value": summary["total"], "color": "0D3349"},
        {"label": "✅ Passed", "value": summary["passed"], "color": "4CAF50"},
        {"label": "❌ Failed", "value": summary["failed"], "color": "F44336"},
        {"label": "⏭ Skipped", "value": summary["skipped"], "color": "FFC107"}
    ]
    
    ws.row_dimensions[8].height = 20
    ws.row_dimensions[9].height = 20

    for col_idx, s in enumerate(stats, start=1):
        ws.merge_cells(start_row=8, start_column=col_idx, end_row=9, end_column=col_idx)
        cell = ws.cell(row=8, column=col_idx)
        cell.value = f"{s['label']}\n{s['value']}"
        cell.font = Font(name=FONT_NAME, size=14, bold=True, color="FFFFFF")
        cell.fill = PatternFill(start_color=s["color"], end_color=s["color"], fill_type="solid")
        cell.alignment = Alignment(horizontal="center", vertical="center", wrap_text=True)

    pass_rate = (summary["passed"] / summary["total"] * 100) if summary["total"] > 0 else 0
    ws.cell(row=11, column=1, value=f"Pass Rate: {pass_rate:.1f}%").font = Font(name=FONT_NAME, size=12, bold=True, color="2E7D32")
    ws.row_dimensions[11].height = 20

    # Category breakdown table header
    headers = ["Category", "Total", "Passed", "Failed", "Pass %"]
    header_row = 13
    for col_idx, h in enumerate(headers, start=1):
        c = ws.cell(row=header_row, column=col_idx, value=h)
        c.font = FONT_HEADER
        c.fill = PatternFill(start_color=COLORS["HEADER_BG"], end_color=COLORS["HEADER_BG"], fill_type="solid")
        c.alignment = Alignment(horizontal="center")
    ws.row_dimensions[header_row].height = 20

    # Calculate categories
    by_category = {}
    for r in results:
        cat = r["category"]
        if cat not in by_category:
            by_category[cat] = {"total": 0, "passed": 0, "failed": 0}
        by_category[cat]["total"] += 1
        if r["status"] == "PASS":
            by_category[cat]["passed"] += 1
        elif r["status"] == "FAIL":
            by_category[cat]["failed"] += 1

    current_row = 14
    for idx, (cat, data) in enumerate(by_category.items()):
        pct = (data["passed"] / data["total"] * 100) if data["total"] > 0 else 0
        row_values = [cat, data["total"], data["passed"], data["failed"], f"{pct:.0f}%"]
        
        for col_idx, val in enumerate(row_values, start=1):
            cell = ws.cell(row=current_row, column=col_idx, value=val)
            cell.font = FONT_BODY
            cell.alignment = Alignment(horizontal="center" if col_idx > 1 else "left")
            
            # Alt row shading
            if idx % 2 == 0:
                cell.fill = PatternFill(start_color=COLORS["ALT_ROW"], end_color=COLORS["ALT_ROW"], fill_type="solid")
            
            if col_idx == 3:
                cell.font = FONT_PASS
            elif col_idx == 4:
                cell.font = FONT_FAIL
        ws.row_dimensions[current_row].height = 18
        current_row += 1

def _build_details_sheet(wb, results):
    ws = wb.create_sheet(title="📝 Test Details")
    ws.sheet_properties.tabColor = "4CAF50"

    headers = [
        ("Test ID", 10),
        ("Category", 25),
        ("Test Name", 50),
        ("Type", 18),
        ("Status", 12),
        ("Duration (s)", 15),
        ("Error / Notes", 55)
    ]
    
    # Header Row
    ws.row_dimensions[1].height = 24
    for col_idx, (h, w) in enumerate(headers, start=1):
        cell = ws.cell(row=1, column=col_idx, value=h)
        cell.font = FONT_HEADER
        cell.fill = PatternFill(start_color=COLORS["HEADER_BG"], end_color=COLORS["HEADER_BG"], fill_type="solid")
        cell.alignment = Alignment(horizontal="center", vertical="center")
        col_letter = get_column_letter(col_idx)
        ws.column_dimensions[col_letter].width = w

    # Rows
    thin_border = Border(
        left=Side(style='thin', color="DDDDDD"),
        right=Side(style='thin', color="DDDDDD"),
        top=Side(style='thin', color="DDDDDD"),
        bottom=Side(style='thin', color="DDDDDD")
    )

    for idx, r in enumerate(results):
        row_num = idx + 2
        dur_val = f"{(r['duration'] / 1000):.2f}" if r["duration"] != "-" and isinstance(r["duration"], (int, float)) else "-"
        
        ws.cell(row=row_num, column=1, value=r["id"]).alignment = Alignment(horizontal="center")
        ws.cell(row=row_num, column=2, value=r["category"])
        ws.cell(row=row_num, column=3, value=r["name"])
        ws.cell(row=row_num, column=4, value=r.get("type", "Functional"))
        
        sc = ws.cell(row=row_num, column=5, value=r["status"])
        sc.alignment = Alignment(horizontal="center")
        if r["status"] == "PASS":
            sc.fill = PatternFill(start_color="BBDEFB", end_color="BBDEFB", fill_type="solid")
            sc.font = FONT_PASS
        elif r["status"] == "FAIL":
            sc.fill = PatternFill(start_color="FFCDD2", end_color="FFCDD2", fill_type="solid")
            sc.font = FONT_FAIL
        else:
            sc.fill = PatternFill(start_color="FFE0B2", end_color="FFE0B2", fill_type="solid")
            sc.font = FONT_BOLD

        ws.cell(row=row_num, column=6, value=dur_val).alignment = Alignment(horizontal="center")
        
        err_cell = ws.cell(row=row_num, column=7, value=r.get("error", ""))
        err_cell.alignment = Alignment(wrap_text=True)

        for col_idx in range(1, 8):
            cell = ws.cell(row=row_num, column=col_idx)
            cell.border = thin_border
            if cell.font == FONT_BODY or cell.font is None:
                cell.font = FONT_BODY
            
            # Alt row shading
            if idx % 2 == 0 and r["status"] != "PASS" and r["status"] != "FAIL":
                cell.fill = PatternFill(start_color=COLORS["ALT_ROW"], end_color=COLORS["ALT_ROW"], fill_type="solid")
            elif idx % 2 == 0 and col_idx != 5:
                cell.fill = PatternFill(start_color=COLORS["ALT_ROW"], end_color=COLORS["ALT_ROW"], fill_type="solid")

        ws.row_dimensions[row_num].height = 20

    ws.auto_filter.ref = f"A1:G{len(results) + 1}"
    ws.freeze_panes = "A2"

def _build_category_sheet(wb, results):
    ws = wb.create_sheet(title="📊 By Category")
    ws.sheet_properties.tabColor = "E94560"

    headers = [
        ("Category", 30),
        ("Test Type", 20),
        ("Test Name", 50),
        ("Status", 12)
    ]
    
    # Header Row
    ws.row_dimensions[1].height = 22
    for col_idx, (h, w) in enumerate(headers, start=1):
        cell = ws.cell(row=1, column=col_idx, value=h)
        cell.font = FONT_HEADER
        cell.fill = PatternFill(start_color=COLORS["HEADER_BG"], end_color=COLORS["HEADER_BG"], fill_type="solid")
        cell.alignment = Alignment(horizontal="center", vertical="center")
        col_letter = get_column_letter(col_idx)
        ws.column_dimensions[col_letter].width = w

    # Sorted Results
    sorted_results = sorted(results, key=lambda x: (x["category"], x.get("type", "Functional"), x["id"]))

    for idx, r in enumerate(sorted_results):
        row_num = idx + 2
        ws.cell(row=row_num, column=1, value=r["category"])
        ws.cell(row=row_num, column=2, value=r.get("type", "Functional"))
        ws.cell(row=row_num, column=3, value=r["name"])
        
        sc = ws.cell(row=row_num, column=4, value=r["status"])
        sc.alignment = Alignment(horizontal="center")
        if r["status"] == "PASS":
            sc.fill = PatternFill(start_color="BBDEFB", end_color="BBDEFB", fill_type="solid")
            sc.font = FONT_PASS
        elif r["status"] == "FAIL":
            sc.fill = PatternFill(start_color="FFCDD2", end_color="FFCDD2", fill_type="solid")
            sc.font = FONT_FAIL

        for col_idx in range(1, 5):
            cell = ws.cell(row=row_num, column=col_idx)
            if cell.font == FONT_BODY or cell.font is None:
                cell.font = FONT_BODY
            
            # Alt row shading
            if idx % 2 == 0 and col_idx != 4:
                cell.fill = PatternFill(start_color=COLORS["ALT_ROW"], end_color=COLORS["ALT_ROW"], fill_type="solid")
        ws.row_dimensions[row_num].height = 18

    ws.auto_filter.ref = f"A1:D{len(results) + 1}"
    ws.freeze_panes = "A2"
